CAPI 1.11

martiniadam opened the pull request:
This PR updates CAPH to use CAPI 1.11, and updates the provider to use v1beta2 CAPI API's

Please note that this will bring up CAPH to v1.2.0

martiniadam committed:
move to v1beta1 and CAPI 1.11

martiniadam committed:
update controllers to use v1beta2

martiniadam committed:
Update everything else

martiniadam committed:
update kustomize

martiniadam committed:
update manager_image_patch

martiniadam committed:
update v1beta2 metadata

martiniadam committed:
update to 1.11.2

martiniadam committed:
updates to images/versions

martiniadam committed:
convert from v1beta1 to v1beta2

martiniadam committed:
update auth_proxy

martiniadam committed:
fix verbosity

martiniadam committed:
populate all Reason fields for Conditions

martiniadam committed:
update the api to use custom initialization statuses

martiniadam committed:
fix to condition

martiniadam committed:
add tag

martiniadam committed:
add defaults for ports

martiniadam committed:
chore: upgrade CAPI to v1.11.5

martiniadam committed:
fix: change ManagedDisk to pointer type with omitempty to fix v1beta2 validation

The v1beta2 OSDisk.ManagedDisk field was a value type which caused it to serialize
as {"storageAccountType":""} when empty, failing validation with error:
  spec.template.spec.osDisk.managedDisk: Required value

Changed ManagedDisk from value type to pointer type with omitempty tag so empty
values are properly omitted from serialization. Added manual conversion functions
to handle the type difference between v1beta1 (value) and v1beta2 (pointer).

martiniadam committed:
fix: change Image, OSDisk, AvailabilityZone to pointer types with omitempty

This comprehensive fix addresses v1beta2 validation issues where empty nested structs
with required fields would fail CRD validation. By changing these from value types
to pointer types with omitempty, empty values are omitted from serialization entirely.

Changes:
- AzureStackHCIMachineSpec: Image, OSDisk, AvailabilityZone now pointers
- AzureStackHCIVirtualMachineSpec: Image, OSDisk, AvailabilityZone now pointers
- AzureStackHCILoadBalancerSpec: Image now pointer
- ManagedDisk in OSDisk: now pointer (previous fix)
- Added conversion functions for all affected types
- Updated controllers to handle pointer types correctly

martiniadam committed:
Fix v1beta1 to v1beta2 conversion for Initialization.Provisioned field

During upgrade from CAPI 1.10 (v1beta1) to CAPI 1.11 (v1beta2), the
AzureStackHCICluster and AzureStackHCIMachine status conversion was not
populating the Initialization.Provisioned field from the v1beta1 Ready field.

This caused the v1beta2 machine controller to always return early with
'Cluster infrastructure is not ready yet' because it checks
Status.Initialization.Provisioned before reconciling machines.

Fix: Convert v1beta1 Ready=true to v1beta2 Initialization.Provisioned=true
for both cluster and machine status.

martiniadam committed:
Delay InfrastructureProvisioned until ControlPlaneEndpoint is valid

This preserves the timing behavior from CAPI v1beta1 where KCP waited for
InfrastructureReady (set late) rather than InfrastructureProvisioned (set
early in v1beta2).

Without this delay, KCP may adopt standalone control plane machines during
bootstrap and trigger an unwanted rolling update due to KubeadmConfig
mismatches between the standalone KubeadmConfig and KCP's kubeadmConfigSpec.

martiniadam committed:
Revert "Delay InfrastructureProvisioned until ControlPlaneEndpoint is valid"

This reverts commit 0d6371ed1cb195ed1efd99ec65fa40ce28c0d419.

martiniadam committed:
updates for machine ready getting stuck

martiniadam committed:
fix hot reconcile loop: use conditions.Set instead of append for VM conditions

martiniadam commented:
Successful run of KVA pipelines with dev image (as well as cloud-operator, mgmtappl changes) https://dev.azure.com/msazure/msk8s/_build/results?buildId=153302557&view=results

nwoodmsft commented on controllers/azurestackhcicluster_controller.go:
```diff
@@ -157,11 +157,29 @@ func (r *AzureStackHCIClusterReconciler) reconcileNormal(clusterScope *scope.Clu
+			outOfCapacityCondition := metav1.Condition{
+				Type:    infrav1.NetworkInfrastructureReadyCondition,
+				Status:  metav1.ConditionFalse,
+				Reason:  infrav1.OutOfMemoryReason,
```
Why OutOfMemoryReason is returned for OutOfCapacity case?

martiniadam replied:
A mistake, reason is now OutOfCapacity

nwoodmsft commented on controllers/azurestackhcimachine_controller.go:
```diff
@@ -207,11 +210,28 @@ func (r *AzureStackHCIMachineReconciler) reconcileNormal(machineScope *scope.Mac
 		return reconcile.Result{}, nil
 	}
 
+	// Set machine infrastructure as provisioned so CAPI knows we're ready to accept bootstrap data
```
On master, Status.Ready was only set after VMStateSucceeded (old line 253). The new code at lines 221-231 would set Initialization.Provisioned = true before the VM is even created by MOC, along with an
  immediate PatchObject() to persist it. That seems strange?

It looks like Initialization.Provisioned is just v1beta2's equivalent of the old Status.Ready. CAPI should populate 
  Machine.Spec.Bootstrap.DataSecretName independently (the old code checked for it at line 216 with no dependency on Ready). 

SetReady() is called on line 278 and that should set Provisioned = true correctly...so I am not sure if we should be handling this in the line 226-233 block...

nwoodmsft replied:
Let's also compare to the behavior of the azureedgemachine_controller.go in mono repo (it doesn't seem to set initialization.provisioned until the end of reconcilation, after the bootstrap data is available).

martiniadam replied:
That makes sense, I may have added this as a workaround for some deploy issues I was having before. Fixed and retesting.

nwoodmsft commented on controllers/azurestackhcicluster_controller.go:
```diff
@@ -233,14 +292,21 @@ func (r *AzureStackHCIClusterReconciler) reconcileDelete(clusterScope *scope.Clu
 		wrappedErr := errors.Wrapf(err, "error deleting AzureStackHCICluster %s/%s", azureStackHCICluster.Namespace, azureStackHCICluster.Name)
 		r.Recorder.Eventf(azureStackHCICluster, corev1.EventTypeWarning, "FailureClusterDelete", wrappedErr.Error())
-		conditions.MarkFalse(azureStackHCICluster, infrav1.NetworkInfrastructureReadyCondition, clusterv1.DeletionFailedReason, clusterv1.ConditionSeverityWarning, err.Error())
+		conditions.Set(azureStackHCICluster, metav1.Condition{
```
Missing Status field in metav1.Condition ?

metav1.Condition.Status is +required with +kubebuilder:validation:Enum=True;False;Unknown — the Kubernetes API definition explicitly marks it as required and validates it must be one of those three values. An
   empty string "" is invalid.

All 4 other "DeletionFailed" conditions in this file correctly include Status: metav1.ConditionFalse

martiniadam replied:
Yes thank you, fixed.

martiniadam committed:
Addressing PR feedback

martiniadam committed:
Fix: remove premature Provisioned=true, add nil guards for pointer fields

- Remove early Initialization.Provisioned=true in machine controller that
  violated v1beta2 contract by marking machines as provisioned before VM
  creation. Provisioned is now only set via SetReady() after VMStateSucceeded.
- Add nil guard for LoadBalancer Image.Version access
- Add nil guards for AvailabilityZone pointer access in machine and VM scopes

martiniadam commented:
Successful run after review changes (one test flake) https://msazure.visualstudio.com/msk8s/_build/results?buildId=153681433&view=results

martiniadam committed:
Fix conversionReviewVersions: use v1beta1 instead of invalid v1beta2

conversionReviewVersions refers to the apiextensions.k8s.io ConversionReview
API version, not the CAPI API group version. The only valid versions are v1
and v1beta1. v1beta2 does not exist for ConversionReview.

martiniadam committed:
Update CAPI_VERSION from v1.11.1 to v1.11.5 in Makefile

martiniadam committed:
Revert config/default to master (remove dev image and fix webhook patch)

martiniadam committed:
Fix admissionReviewVersions: v1beta2 -> v1beta1, add trailing newline to metadata.yaml

admissionReviewVersions refers to the Kubernetes AdmissionReview API version
(admissionregistration.k8s.io), not the CAPI contract version. Only v1 and
v1beta1 exist for AdmissionReview.
